packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -3643,14 +3643,28 @@ class _UpdateTextSelectionAction<T extends DirectionalCaretMovementIntent> exten
       );
     }
 
-    final TextPosition extent = textBoundarySelection.extent;
+    late final TextPosition position;
+    if (!intent.expand) {
+      position = textBoundarySelection.extent;
+    } else {
+      final bool inOrder = textBoundarySelection.baseOffset <= textBoundarySelection.extentOffset;
+      final bool towardsExtent = intent.forward == inOrder;
+      position = towardsExtent
+          ? textBoundarySelection.extent
+          : textBoundarySelection.base;
+    }
     final TextPosition newExtent = intent.forward
-      ? textBoundary.getTrailingTextBoundaryAt(extent)
-      : textBoundary.getLeadingTextBoundaryAt(extent);
+      ? textBoundary.getTrailingTextBoundaryAt(position)
+      : textBoundary.getLeadingTextBoundaryAt(position);
 
-    final TextSelection newSelection = collapseSelection
-      ? TextSelection.fromPosition(newExtent)
-      : textBoundarySelection.extendTo(newExtent);
+    late final TextSelection newSelection;
```

Comment:
unnecessary `late`?

Reply:
Undone with refactoring to use a new Intent.